packages/block-editor/src/components/rich-text/index.js

```diff
@@ -39,6 +39,7 @@ import {
 } from '@wordpress/rich-text';
 import deprecated from '@wordpress/deprecated';
 import { isURL } from '@wordpress/url';
+import { regexp } from '@wordpress/shortcode';
```

Comment:
I didn't see this code at the time of approval. I feel like this belongs in the paste handler together with the rest of the short code logic?

Reply:
I think code may have changed since your approval and Miguel's. I didn't think of pinging you for your re-approval after Miguel's requested changes, sorry.

However, I don't know how we can move this logic to the paste handler. The trick here is to switch to `BLOCKS` mode if 1) we're in an empty line and 2) the pasted text is a shortcode. This logic requires knowledge of the rich text value and the pasted content. My thinking was that having this on the rich-text component would make sense as we already do the same for URLs. How does this look? Happy to refactor if another approach is preferred.

Reply:
Also the dependencies are missing in the package file

Reply:
> Also the dependencies are missing in the package file

Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Reply:
Ouch :disappointed: Here's the fix for the missing dependency https://github.com/WordPress/gutenberg/pull/22086

Not sure I've done this before so I may have missed something.

Reply:
Do we need to make a package release due to this? I guess we do unless we can wait until next week's release. Not sure how you managed previous precedents if any, but happy to do it myself as a future reminder of this kind of thing.

Reply:
>Do we need to make a package release due to this?

I think as long as these updates haven't been included in a packages release, there's still time to correct the mistake without needing to do any extra publishing.

As best I can tell, the latest version doesn't include this line of code:

https://unpkg.com/browse/@wordpress/block-editor@3.10.0/build-module/components/rich-text/index.js

Reply:
oh, nice

Reply:
> > Also the dependencies are missing in the package file
> 
> Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Something must not be working correctly, but we do define `import/no-extraneous-dependencies` as a rule, which is intended to enforce this:

https://github.com/WordPress/gutenberg/blob/9db5982eda68e8f923e60792242702c8a5960822/.eslintrc.js#L133

>Forbid the import of external modules that are not declared in the package.json's dependencies, devDependencies, optionalDependencies, peerDependencies, or bundledDependencies. 

https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-extraneous-dependencies.md

Reply:
Very strange indeed. It fails locally!

```
 ⇒ npm run lint-js                                                        

> gutenberg@8.0.0 lint-js /Users/andrew/Documents/Code/gutenberg
> wp-scripts lint-js


/Users/andrew/Documents/Code/gutenberg/packages/block-editor/src/components/rich-text/index.js
  42:1  error  '@wordpress/shortcode' should be listed in the project's dependencies. Run 'npm i -S @wordpress/shortcode' to add it  import/no-extraneous-dependencies

✖ 1 problem (1 error, 0 warnings)
```

Reply:
The [Travis job](https://travis-ci.com/github/WordPress/gutenberg/jobs/324466082) for this PR didn't report anything neither for all subsequent PRs. Just looked at the logs and there aren't any messages about the failure either (it could be that the exit code was wrong for some reason but the message was there).

I can confirm that `npm run lint-js` and `npm run lint` do fail locally on `master` prior to the merge of https://github.com/WordPress/gutenberg/pull/22086 Puzzling.

Reply:
> Aside: I thought we had an ESLint rule which verified that any referenced dependencies should be listed in `package.json` 🤔

Oh, that's unfortunate, I let my guard down too. 😞 

> I didn't see this code at the time of approval. I feel like this belongs in the paste handler together with the rest of the short code logic?

Perhaps, but there are a lot of symmetries with what we do with pasted URLs and embeds. Whatever we do, we should do for both shortcodes and embeds. Perhaps that means letting `pasteHandler` choose whether to force `mode: 'BLOCKS'` when initial `mode` is `'AUTO'`; perhaps that means exposing a utility like `inferMode` or `shouldForceBlocksMode` from `@wordpress/blocks` that RichText can use.

Reply:
> Very strange indeed. It fails locally!

After some more debugging, I think I've figured it out.

Fix (and debugging details) at: #22088

Reply:
Sorry, I don't mind the code, it's something we can always easily change. I was thinking we could also pass empty state to the raw handler. I just thought it might be nice to keep shortcode dependencies in the raw handler since it already depends on it and RichText doesn't. I just noticed the code when I saw the package error.